File: src/Utils/IntelliTimeStampGenerator.cpp
//
// Created by tony on 06/01/24.
//
#include <Utils/IntelliTimeStampGenerator.h>

bool INTELLI::IntelliTimeStampGenerator::setConfig(INTELLI::ConfigMapPtr cfg) {
  cfgGlobal = cfg;
  eventRateTps = cfg->tryI64("eventRateTps", 100, true);
  timeStepUs = cfg->tryI64("timeStepUs", 100, true);
  timeStamper_zipfEvent = cfg->tryI64("timeStamper_zipfEvent", 0, true);
  timeStamper_zipfEventFactor = cfg->tryDouble("timeStamper_zipfEventFactor", 0.1, true);
  testSize = cfg->tryI64("streamingTupleCnt", 0, true);
  staticDataSet = cfg->tryI64("staticDataSet", 0, true);
  md.setSeed(seed);
  generateEvent();
  generateArrival();
  generateFinal();
  return true;
}

void INTELLI::IntelliTimeStampGenerator::generateEvent() {
  uint64_t maxTime = testSize * 1000 * 1000 / eventRateTps;
  if (staticDataSet) {
    eventS.resize(testSize);
    eventS.assign(testSize, 0);// Create vector of size 'size' with all elements set to 0
  } else if (timeStamper_zipfEvent) {
    INTELLI_INFO("Use zipf for event time, factor=" + to_string(timeStamper_zipfEventFactor));
    INTELLI_INFO("maxTime=" + to_string(maxTime) + "us" + "rate=" + to_string(eventRateTps) + "K, cnt=" +
        to_string(testSize));
    eventS =
        md.genZipfTimeStamp<uint64_t>(testSize, maxTime,
                                      timeStamper_zipfEventFactor);
  } else {
    // uint64_t tsGrow = 1000 * timeStepUs / eventRateKTps;
    eventS = md.genSmoothTimeStamp(testSize, maxTime);
  }
  INTELLI_INFO("Finish the generation of event time");
}

void INTELLI::IntelliTimeStampGenerator::generateArrival() {

  INTELLI_INFO("Finish the generation of arrival time");
}

std::vector<INTELLI::IntelliTimeStampPtr> INTELLI::IntelliTimeStampGenerator::constructTimeStamps(
    std::vector<uint64_t> _eventS,
    std::vector<uint64_t> _arrivalS) {
  size_t len = _eventS.size();
  std::vector<INTELLI::IntelliTimeStampPtr> ru = std::vector<INTELLI::IntelliTimeStampPtr>(len);
  for (size_t i = 0; i < len; i++) {
    ru[i] = newIntelliTimeStamp(_eventS[i], _arrivalS[i], 0);
  }
  return ru;
}

void INTELLI::IntelliTimeStampGenerator::generateFinal() {
  myTs = constructTimeStamps(eventS, eventS);
}
std::vector<INTELLI::IntelliTimeStampPtr> INTELLI::IntelliTimeStampGenerator::getTimeStamps() {
  return myTs;
}

File: benchmark/scripts/DAGNN/intelli_timestamp_generator.py
from typing import List
import random
class IntelliTimeStamp:
    def __init__(self, eventTime=0, arrivalTime=0, processedTime=0):
        self.eventTime = eventTime
        self.arrivalTime = arrivalTime
        self.processedTime = processedTime
    
    def __repr__(self):
        return f"IntelliTimeStamp(eventTime={self.eventTime}, arrivalTime={self.arrivalTime}, processedTime={self.processedTime}"

def generate_random_list(length: int, maxTime: int, minTime=0) -> List[int]:
    return [random.randint(0, maxTime - 1) for _ in range(length)]

def gen_smooth_timestamp(length: int, maxTime: int) -> List[int]:
    ret = generate_random_list(length, maxTime)
    ret.sort()  # Incremental rearrangement (sort the list)
    return ret

class IntelliTimeStampGenerator:
    def __init__(self, testSize=0):
        self.testSize = testSize
        self.eventS: List[int] = []
        self.arrivalS: List[int] = []
        self.eventRateTps = 100
        self.timeStepUs = 40
        self.seed = 114514
        self.staticDataSet = 0
        self.myTs: List[IntelliTimeStamp] = []
        self.generateEvent()
        self.generateArrival()
        self.generateFinal()

    def generateEvent(self):
        # Implementation for generating the vector of events
        
        maxTime = self.testSize * 1000 * 1000 // self.eventRateTps
        self.eventS = gen_smooth_timestamp(self.testSize, maxTime);  # Create a list of size 'testSize'
        print("Finish the generation of event time")
        return

    def generateArrival(self):
        # Implementation for generating the vector of arrivals
        pass

    def generateFinal(self):
        # Implementation for generating the final result of s and r
        self.myTs=self.constructTimeStamps(self.eventS, self.eventS)
        return

    def constructTimeStamps(self, eventS: List[int], arrivalS: List[int]) -> List[IntelliTimeStamp]:
        
        ru = [IntelliTimeStamp(event, arrival, 0) for event, arrival in zip(eventS, arrivalS)]
        return ru

    def setConfig(self, cfg: dict) -> bool:
        # Implementation for setting the global config map related to this TimerStamper
        return True

    def getTimeStamps(self) -> List[IntelliTimeStamp]:
        # Implementation for getting the vector of time stamps
        return self.myTs
    
tsg = IntelliTimeStampGenerator(50)
print(tsg.arrivalS)
print(tsg.eventS)
print(tsg.myTs)